server/src/copilot/utils/embeddingsUtils.ts

```diff
@@ -0,0 +1,162 @@
+import * as math from 'mathjs';
+import { Documents, ModelEmbeddings, ModelsEmbeddingsData, TemplateEmbeddings } from './types';
+import { log } from 'console';
+
+function cosineSimilarity(vec1: number[], vec2: number[]): number {
+    const dotProduct = math.dot(vec1, vec2);
+    const normVec1:any = math.norm(vec1);
+    const normVec2:any = math.norm(vec2);
+    return dotProduct / (normVec1 * normVec2);
+}
+
+export function fetchRelevantNamespaces(modelNamespaces: { [key: string]: ModelEmbeddings }, promptEmbedding: number[], provider: string, topN = 4): string[] {
+    log('Fetching relevant namespaces' + promptEmbedding );
+    const namespaceSimilarities: [number, string, string][] = [];
+
+    for (const [file, obj] of Object.entries(modelNamespaces)) {
+        let embeddings: number[] | undefined; 
+
+        if (provider === 'gemini' && obj.gemini?.embeddings?.embedding) {
+            embeddings = obj.gemini.embeddings.embedding;
+        } else if (provider === 'openai' && obj.openai?.embeddings) {
+            embeddings = obj.openai.embeddings;
+        }
+
+        if (embeddings && Array.isArray(embeddings)) {
+            const similarity = cosineSimilarity(promptEmbedding, embeddings);
+            namespaceSimilarities.push([similarity, file, obj.fileContent]);
+        }
+    }
+
+    const topNamespaces = namespaceSimilarities.sort((a, b) => b[0] - a[0]).slice(0, topN);
+
+    const relevantNamespaces = topNamespaces.map(([similarity, file, fileContent]) => {
+        return `${file}\n`;
+    });
+
+    return relevantNamespaces;
+}
+
+export function fetchRelevantTemplates(templateEmbeddings: TemplateEmbeddings, promptEmbedding: number[], provider: string, topN = 4): string[] {
+    log('Fetching relevant templates' + promptEmbedding);
+    const templateSimilarities: [number, string, string, string][] = [];
+
+    for (const [templateName, template] of Object.entries(templateEmbeddings)) {
+        let embeddings: number[] | undefined;
+
+        if (provider === 'gemini' && template.grammar?.embeddings?.gemini?.embedding) {
```

Comment:
Again, we want to move any LLM specific functionality into an interface that is implemented for that LLM, rather than have lots of `provider ===` scattered around.

Reply:
Fixed, now in interface LargeLanguageModel we have getDocsEmbeddings functionality which provide embeddings for specific provider directly.